sdlib/dmd/gc.d

```diff
@@ -0,0 +1,105 @@
+module dmd.gc;
+
+import d.gc.tcache;
+import d.gc.spec;
+import d.gc.slab;
+import d.gc.emap;
+
+extern(C):
+
+// Some of the bits from druntime.
+enum BlkAttr : uint {
+	FINALIZE = 0b0000_0001,
+	NO_SCAN = 0b0000_0010,
+	APPENDABLE = 0b0000_1000
+}
+
+// ptr: the pointer to query
+// base: => the true base address of the block
+// size: => the full usable size of the block
+// pd: => the page descriptor (used internally later)
+// flags: => what the flags of the block are
+//
+bool __sd_gc_fetch_alloc_info(void* ptr, void** base, size_t* size,
+                              PageDescriptor* pd, BlkAttr* flags) {
```

Comment:
Is it expected for druntime to be able to read the page descriptor? This is an internal format for the GC and will be subject to changes. Also why is base a `void**` rather than a `void*` ?

Reply:
> Is it expected for druntime to be able to read the page descriptor?

It's technically not important at the moment on druntime side. At this moment, the only thing I use it for is to check if it's a large block or not. This goes into the decision of whether to reduce the array size by 1 to avoid the adjacent block issue we talked about.

Other than that, it's basically "private data" to give back to SDC so it can look up the block info. See the `__sd_gc_get_array_used` function and the accompanying set function. Those use the PageDescriptor to figure things out.

I could just change it to a size_t, as it is on the other side, and use reinterpret casting? I didn't think having the type defined here would be a problem.

> Also why is base a `void**` rather than a `void*` ?

It is an output parameter.

the input parameter is the `ptr`, which could be an interior pointer. The `base` is set to the true start of the slab/extent.

Reply:
Fair enough. I'm not sure I'm such a fan of that solution, but it'll work and doesn't have a widespread impact, so it's ok.

Reply:
I'm OK with another solution if you can think of one. I'm not even entirely sure this is the *most optimal* solution. But it does work, and gets us to a working array runtime.

For the large blocks, I think it can't be helped to store the extent pointer. I don't think it's worth doing a lookup for each append in the emap. So we need at least that.